mapiproxy/servers/default/nspi/dcesrv_exchange_nsp.c

```diff
@@ -325,30 +298,88 @@ static void dcesrv_NspiUpdateStat(struct dcesrv_call_state *dce_call, TALLOC_CTX
 		}
 	}
 
-	if (-r->in.pStat->Delta > row) {
-		row = 0;
-		r->in.pStat->CurrentRec = mids->aulPropTag[row];
-	}
-	else if (r->in.pStat->Delta + row >= row_max) {
-		row = row_max;
-		r->in.pStat->CurrentRec = MID_END_OF_TABLE;
+	if (r->in.pStat->Delta != 0) {
+		/* Adjust row  by Delta */
+		if (r->in.pStat->Delta > 0) {
+			row += r->in.pStat->Delta;
+			if (row > row_max) {
+				row = row_max;
+			}
+		} else {
+			if (abs(r->in.pStat->Delta) >= row) {
+				row = 0;
+			} else {
+				row -= abs(r->in.pStat->Delta);
+			}
+		}
 	}
-	else {
-		row += r->in.pStat->Delta;
-		r->in.pStat->CurrentRec = mids->aulPropTag[row];
+	
+	if (row == row_max) {
+		/* even if row = 0 we should return MID_END_OF_TABLE */
+		r->out.pStat->CurrentRec = MID_END_OF_TABLE;
+        } else if (row == 0) {
+		r->out.pStat->CurrentRec = MID_BEGINNING_OF_TABLE;
+	} else {
+		r->out.pStat->CurrentRec = mids->aulPropTag[row];
 	}
 
-	r->in.pStat->Delta = 0;
-	r->in.pStat->NumPos = row;
-	r->in.pStat->TotalRecs = row_max;
+	r->out.pStat->Delta = 0;
+	r->out.pStat->NumPos = row;
+	r->out.pStat->TotalRecs = row_max + 1;
+	r->out.plDelta = r->in.plDelta;
+	if (r->in.plDelta != NULL) {
+		*(r->out.plDelta) = r->out.pStat->NumPos - r->in.pStat->NumPos;
+	}
 
 end:
-	r->out.pStat = r->in.pStat;
-
 	DCESRV_NSP_RETURN(r, retval, local_mem_ctx);
 }
 
 /**
+   \details exchange_nsp NspiUpdateStat (0x2) function
+
+   \param dce_call pointer to the session context
+   \param mem_ctx pointer to the memory context
+   \param r pointer to the NspiUpdateStat request data
+
+   \return MAPI_E_SUCCESS on success
+*/
+/* TODO:
+ - Use SortType. sortLocale and ContainerId
+- Check CodePage first
+ */
+static void dcesrv_NspiUpdateStat(struct dcesrv_call_state *dce_call, TALLOC_CTX *mem_ctx, struct NspiUpdateStat *r)
+{
+	struct emsabp_context		*emsabp_ctx = NULL;
+	bool				container_exists;
+
+	DEBUG(3, ("exchange_nsp: NspiUpdateStat (0x2)"));
+	/* Ensure incoming user is authenticated */
+	if (!dcesrv_call_authenticated(dce_call)) {
+		DEBUG(1, ("No challenge requested by client, cannot authenticate\n"));
+		DCESRV_NSP_RETURN(r, MAPI_E_LOGON_FAILED, NULL);
+	}
+
+	DCESRV_NSP_RETURN_IF(r->in.pStat->CodePage == CP_UNICODE, r, MAPI_E_NO_SUPPORT, NULL);
+
+	emsabp_ctx = dcesrv_find_emsabp_context(&r->in.handle->uuid);
+	if (!emsabp_ctx) {
+		DCESRV_NSP_RETURN(r, MAPI_E_CALL_FAILED, NULL);
```

Comment:
This could be simplified to `DCESRV_NSP_RETURN_IF(!emsabp_ctx, r, MAPI_E_CALL_FAILED, NULL);`